diff --git a/lib/Transforms/Utils/MetaRenamer.cpp b/lib/Transforms/Utils/MetaRenamer.cpp
index 233bc12..c53aa2f 100644
--- a/lib/Transforms/Utils/MetaRenamer.cpp
+++ b/lib/Transforms/Utils/MetaRenamer.cpp
@@ -96,7 +96,10 @@ namespace {
       // Rename all functions
       for (Module::iterator FI = M.begin(), FE = M.end();
            FI != FE; ++FI) {
-        FI->setName(metaNames[prng.rand() % array_lengthof(metaNames)]);
+        if ((FI->getCallingConv() == llvm::CallingConv::SPIR_FUNC && !FI->isDeclaration())
+            || (FI->getCallingConv() != llvm::CallingConv::SPIR_FUNC && FI->getCallingConv() != llvm::CallingConv::SPIR_KERNEL))
+          FI->setName(metaNames[prng.rand() & array_lengthof(metaNames)]);
+
         runOnFunction(*FI);
       }
       return true;
